### ISSUE TITLE
dart2js: error when creating InvocationMirror (wrong kind)

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#23828]
We seem to have a bug in how we generate the invocation mirrors for noSuchMethod when:
* the method name starts with `get`
* for some reason we don't have the reflectiveName available

(basically when we hit [this codepath](https://github.com/dart-lang/sdk/blob/d7c2f4550b64a460acc9785eba14a16a66a44707/pkg/compiler/lib/src/js_emitter/full_emitter/nsm_emitter.dart#L212))

The check for whether the [shortName is a getter prefix](https://github.com/dart-lang/sdk/blob/d7c2f4550b64a460acc9785eba14a16a66a44707/pkg/compiler/lib/src/js_emitter/full_emitter/nsm_emitter.dart#L240) makes us mark the method as a getter (`._kind == 1`). So, even though the arguments are populated correctly internally (`._arguments` is correct) accessing the positionalArguments results empty.

Here is a simple repro:
```dart
import 'package:expect/expect.dart';

confuse(x) { return x; return x; }

class MA {
  noSuchMethod(i) => Expect.equals(i.positionalArguments.length, 1);
}

main() {
  //print(#getFoo);
  confuse(new MA()).getFoo('a');
}
```

The length ends up being 0, if you comment out the print statement in main, or if you change `getFoo` to `foo`, then the length is 1.

**Comment by sigmundch** [comment]
A bisect reveals that this broke recently in dd41983

I'll keep digging in, but @herhut-ggl @floitschG let me know if already know the source of the issue.

**Comment by sigmundch** [comment]
While that change is not modifying the behavior of how we interpret shortNames, it seems that the change added new names to the list that weren't included before.

For the same program, here is the diff in the list of shortNames/longNames before and after the change:

```diff
--- before	2015-07-13 13:09:57.931389529 -0700
+++ after	2015-07-13 13:09:56.815388466 -0700
@@ -1,26 +1,49 @@
   shortNames:
+  '$add',
+  '$gt',
+  '$index',
+  '$lt',
+  '$sub',
+  '',
   'call$0',
   'call$1',
   'call$2',
   'call$3',
+  'forEach$1',
   'get$_contents',
   'get$_name',
   'get$hashMapCellKey',
   'get$hashMapCellValue',
+  'get$length',
   'get$memberName',
   'get$namedArguments',
   'get$positionalArguments',
+  'getFoo$1',
+  'matchAsPrefix$2',
   'set$_contents',
   'set$hashMapCellValue',
+  'startsWith$1',
+  'substring$1',
 
 longNames:
+  '+',
+  '-',
+  '<',
+  '>',
+  '[]',
   '_contents',
   '_contents=',
   '_name',
   'call',
+  'forEach',
+  'getFoo',
   'hashMapCellKey',
   'hashMapCellValue',
   'hashMapCellValue=',
+  'length',
+  'matchAsPrefix',
   'memberName',
   'namedArguments',
   'positionalArguments',
+  'startsWith',
+  'substring',
```

**Comment by sigmundch** [comment]
It seems we have a boolean check that is being done inconsistently between the compiler and the runtime. This used to be consistent before https://github.com/dart-lang/sdk/commit/dd41983945b4ccbae5f4670c6235265444a16c31, but now it's inconsistent in cases like `getFoo`.

Before the change `isTrivialNsmHandler` was returning false for `getFoo` because it matched the "getter" pattern test and `type` was 0:

```dart
if (internalName.startsWith(namer.getterPrefix[0])) return type == 1;
```

After the change, the new check:
```dart
if (internalName is GetterName) return type == 1;
```
doesn't apply because `internalName` is not a getter. So the `isTrivialNsmHandler` function eventually returns true further down.

The JS runtime is using a check similar to what we did in `isTrivialNsmHandler` before this change:

```js
for (var j = 0; j < shortNames.length; j++) {
  var type = 0;
  var shortName = shortNames[j];
  if (shortName[0] == "${namer.getterPrefix[0]}") type = 1;
```

Because we used to filter out all names that looked like getters, we didn't hit this issue before. Now that the tests are slightly different, we end up checking `getFoo` here.

I can see two ways of fixing this: make the filtering stronger as it was before, or change the runtime code to improve the test of getters. This proposed CL that does the latter: https://codereview.chromium.org/1232423004/
